docs/rfc/0000-cli-infrastructure.adoc

```diff
@@ -0,0 +1,237 @@
+
+== Overview
+
+This proposal will outline an binary called `rad`, what it will
+provide, and how it can be extended. Following this we provide details
+on the interaction -- or lack there of -- with the peer-to-peer
+daemon.
+
+We also provide a proposal of how any CLI tool will have
+access to any keys, as well as what needs to be completed for
+providing this.
+
+We close out with how the core components will be organised within
+`radicle-link` and `<<radicle-bins>>` as well as providing the first
+iteration of the core set of subcommands.
+
+== One `rad` to Rule Them All
+
+[quote]
+-------
+One ring to rule them all, one ring to find them,
+One ring to bring them all, and in the darkness bind them;
+-------
+-- J.R.R. Tolkein, Fellowship of the Ring
+
+The integral part of this proposal is that there will be one binary,
+`rad`, that will be the foundation for all other subcommands. It will
+define the global parameters and how to extend it.
+
+We will discuss these in the following two sections.
+
+== The Fellowship of the `rad`
+
+[quoute]
+--------
+“I can’t carry it for you, but I can carry you.” – Sam Gamgee
+--------
+-- J.R.R. Tolkien, The Return of the King
+
+To allow the ecosystem of commands to be extended by ourselves and
+others in the community it will be necessary to extend the `rad`
+command with further subcommands. The details of this are outlined in
+RFC-682#CLI, however, we will give a quick overview here for the sake
+of the reader.
+
+Taking inspiration from tools such as `git` and `cargo`, we propose
+that the `rad` binary can be extended by placing another
+binary on the users `$PATH` (or `%PATH%` on Windows). This
+binary MUST follow the naming convention of `rad-*`. To reiterate
+here, these new commands MUST NOT override the core set outlined here
+(nor in further RFCs).
+
+== Global Parameters
+
+Many subcommands will likely want to share a set of options. These
+options, dubbed as "global parameters" here, will be a defined set --
+exported from `radicle-link` for re-use -- that the base `rad` command
+will always receive and pass onto the subcommand being called.
+
+Thus, all subcommands SHALL receive these global parameters as part of
+their interface. We list these parameters here, but it is NOT an
+exhaustive list:
+
+* `verbose`
+* `quiet`
+* `profile := active (default) | <profile-id>`
+
+== Online & Offline
+
+In RFC-682#Process-Orchestration, it was decided that the peer-to-peer
+damon would not be a long-running process. The terminology we will use
+here is that the daemon is *online* if it is running and connected to
+the network. Otherwise, if it is not running then it is *offline*.
+
+This disctinction becomes important when thinking about the
+functionality of `rad` subcommands. The majority of subcommands that
+interact with the storage can be done while the daemon is offline. The
+only action the author can currently think of — that is online — is
+replication of a URN from the network.
```

Comment:
Is that due to how the current "waiting room" is implemented as an online API call to issue a "request" for an URN? In the spirit of the p2p node converged over the state of the storage is there a potential way forward where requesting/replicating a URN from the network could be done through pure manipulation of the storage in contrast to rely on a running node process?

Reply:
Well, fundamentally the point is that you need to actually have the daemon running so that it can replicate from another peer (waiting room or not).

As pointed out in the discussion https://github.com/radicle-dev/radicle-link/pull/698/files#r649924840 there are potentially two steps to the process. We do an offline track of a new URN, and when we're online it finds providers and replicates if it finds any.

So I guess the question is: Are there any online commands or is it all offline and the daemon knows what to do when it comes online via the "queded tasks" that we made while offline?

Reply:
Perhaps. But you may still want to `start` / `stop` / `Status` it =>

Reply:
> But you may still want to start / stop / Status it =>

Right, I was holding off on any specification for daemon control commands, but I think this has convinced me that we want to define them here :) 

It also feels like it should be a requirement that all commands work offline and the daemon can perform the tasks when it's online. The following diagram must commute:
![image](https://user-images.githubusercontent.com/3503528/122535008-087c7280-d01b-11eb-99b4-93c40b8b6d0c.png)
(image created using [quiver](https://q.uiver.app/))